x-pack/protocol/src/main/java/org/elasticsearch/protocol/xpack/ml/job/process/DataCounts.java

```diff
@@ -0,0 +1,414 @@
+            (p) -> TimeUtil.parseTimeField(p, LATEST_RECORD_TIME.getPreferredName()),
+            LATEST_RECORD_TIME,
+            ValueType.VALUE);
+        PARSER.declareField(ConstructingObjectParser.optionalConstructorArg(),
+            (p) -> TimeUtil.parseTimeField(p, LAST_DATA_TIME.getPreferredName()),
+            LAST_DATA_TIME,
+            ValueType.VALUE);
+        PARSER.declareField(ConstructingObjectParser.optionalConstructorArg(),
+            (p) -> TimeUtil.parseTimeField(p, LATEST_EMPTY_BUCKET_TIME.getPreferredName()),
+            LATEST_EMPTY_BUCKET_TIME,
+            ValueType.VALUE);
+        PARSER.declareField(ConstructingObjectParser.optionalConstructorArg(),
+            (p) -> TimeUtil.parseTimeField(p, LATEST_SPARSE_BUCKET_TIME.getPreferredName()),
+            LATEST_SPARSE_BUCKET_TIME,
+            ValueType.VALUE);
+    }
+
+    private final String jobId;
+    private long processedRecordCount;
+    private long processedFieldCount;
+    private long inputBytes;
+    private long inputFieldCount;
+    private long invalidDateCount;
+    private long missingFieldCount;
+    private long outOfOrderTimeStampCount;
+    private long emptyBucketCount;
+    private long sparseBucketCount;
+    private long bucketCount;
+    private Date earliestRecordTimeStamp;
+    private Date latestRecordTimeStamp;
+    private Date lastDataTimeStamp;
+    private Date latestEmptyBucketTimeStamp;
+    private Date latestSparseBucketTimeStamp;
+
+    public DataCounts(String jobId, long processedRecordCount, long processedFieldCount, long inputBytes,
+                      long inputFieldCount, long invalidDateCount, long missingFieldCount, long outOfOrderTimeStampCount,
+                      long emptyBucketCount, long sparseBucketCount, long bucketCount,
+                      Date earliestRecordTimeStamp, Date latestRecordTimeStamp, Date lastDataTimeStamp,
+                      Date latestEmptyBucketTimeStamp, Date latestSparseBucketTimeStamp) {
+        this.jobId = jobId;
+        this.processedRecordCount = processedRecordCount;
+        this.processedFieldCount = processedFieldCount;
+        this.inputBytes = inputBytes;
+        this.inputFieldCount = inputFieldCount;
+        this.invalidDateCount = invalidDateCount;
+        this.missingFieldCount = missingFieldCount;
+        this.outOfOrderTimeStampCount = outOfOrderTimeStampCount;
+        this.emptyBucketCount = emptyBucketCount;
+        this.sparseBucketCount = sparseBucketCount;
+        this.bucketCount = bucketCount;
+        this.latestRecordTimeStamp = latestRecordTimeStamp;
+        this.earliestRecordTimeStamp = earliestRecordTimeStamp;
+        this.lastDataTimeStamp = lastDataTimeStamp;
+        this.latestEmptyBucketTimeStamp = latestEmptyBucketTimeStamp;
+        this.latestSparseBucketTimeStamp = latestSparseBucketTimeStamp;
+    }
+
+    public DataCounts(String jobId) {
+        this.jobId = jobId;
+    }
```

Comment:
I don't think we should have this constructor on the client side, as there's no way to sensibly add the missing values (and no sensible reason why the user would need to).

Reply:
I am going to make it package private as it does allow for some more easily written tests.